### PR TITLE
Profile Memory use on PortNamesWithRTree Test

### PR DESCRIPTION
**Pull request by alex-mckenna**
Profiling the memory usage of the test `PortNamesWithRTree` led to the discoveries that

  * some parts of GHC2Core were not as strict as they should have been
  * The instances of `ShowX` for tuples up to the max size use a lot of memory (compared to using `TemplateHaskell` to generate the instances)

Dealing with these has reduced the memory footprint of running this test by a reasonable amount, which is good for helping prevent CI slowdown / tests being killed due to using too much memory.

### CODE REVIEW

File: clash-ghc/src-ghc/Clash/GHC/Evaluator.hs [hunk 1]
```diff
@@ -30,14 +30,12 @@ import           Data.Bits
 import           Data.Char           (chr,ord)
 import qualified Data.Either         as Either
 import qualified Data.IntMap         as IntMap
-import           Data.Maybe
-  (fromMaybe, mapMaybe, catMaybes)
+import           Data.Maybe (mapMaybe, catMaybes)
```
[the remark this answers is not in the document]
Comment (alex-mckenna) [reply clash-ghc/src-ghc/Clash/GHC/Evaluator.hs]: Indent as in (1) move to a new line and indent like it was or (2) align it with the other import lists.
The import lists in the file are already relatively inconsistent, it wasn't clear what to prefer

[the remark this answers is not in the document]
Comment (martijnbastiaan) [reply clash-ghc/src-ghc/Clash/GHC/Evaluator.hs]: Ah, well it's either

```haskell
import qualified Data.Maybe
  (fromMaybe, ..
```

or 


```haskell
import qualified Data.Maybe        (fromMaybe, ..
```

where the `        ` lines up with the rest.

[the remark this answers is not in the document]
Comment (martijnbastiaan) [reply clash-ghc/src-ghc/Clash/GHC/Evaluator.hs]: But I see `Evaluator.hs` is super inconsistent anyway, so it doesn't really matter I suppose.

File: clash-prelude/src/Clash/CPP.hs [hunk 2]
```diff
@@ -11,7 +11,7 @@ Maintainer :  Christiaan Baaij <christiaan.baaij@gmail.com>
 #ifdef LARGE_TUPLES
 #define MAX_TUPLE_SIZE 62
 #else
-#define MAX_TUPLE_SIZE 12
+#define MAX_TUPLE_SIZE 15
```
Comment (martijnbastiaan) [inline clash-prelude/src/Clash/CPP.hs]: This is mostly here for our own comfort, as compilation gets _slow_ when increasing the tuple size, so I don't really want it to creep up and up.. would it be possible to change the tests, or is that too much trouble?

Comment (alex-mckenna) [reply clash-prelude/src/Clash/CPP.hs]: So the `Generic` and `ShowX` instances went up to 15 in the previous code. I figured changing it here was the smallest change that 

  * kept things working that worked before (some tests need `ShowX` for more than 12-tuples)
  * removed some inaccuracy (instances for tuples larger than `MAX_TUPLE_SIZE`)

The alternative is to change about 6 or so tests, which isn't too bad but seems a bit odd for a PR which conceptually has nothing to do with the affected tests. I could go back and change the tests as part of #910, and then rebase onto that (lowering the limit back to 12)?

**Comment by alex-mckenna** [comment]
There's a weird break on this branch anyway, so I need to go back and see what changes (presumably to `clash-ghc`) caused that

**Comment by alex-mckenna** [comment]
Update: On hold pending #946

**Comment by martijnbastiaan** [comment]
Thanks for debugging the performance issues! I'll merge it once it passes the CI.

**Pull request merged**